bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/IndexPersistenceMgr.java

```diff
@@ -220,7 +220,19 @@ CachedFileInfo getFileInfo(final Long ledger, final byte masterKey[]) throws IOE
                 } else {
                     fi = readFileInfoCache.get(ledger, loader);
                 }
-            } while (!fi.tryRetain());
+                if (!fi.tryRetain()) {
+                    // defensively ensure that dead fileinfo objects don't exist in the
+                    // cache. They shouldn't if refcounting is correct, but if someone
+                    // does a double release, the fileinfo will be cleaned up, while
+                    // remaining in the cache, which could cause a tight loop in this method.
+                    if (writeFileInfoCache.asMap().remove(ledger, fi)
+                        || readFileInfoCache.asMap().remove(ledger, fi)) {
```

Comment:
if writeFileInfoCache#remove returns true we will miss to check readFileInfoCache#remove (the 'or' will not evaluate the second operand)

Reply:
Oops, fixed.